### ISSUE TITLE
Sri Lanka

### ISSUE THREAD

**Issue by tmtmtmtm** [everypolitician/everypolitician-data#121]
Current: http://www.parliament.lk/en/members-of-parliament/directory-of-members/?cletter=A

Past: http://www.parliament.lk/en/members-of-parliament/directory-of-past-members/?cletter=A

**Comment by Abbie2020** [comment]
I'm having a go at writing a scraper for this.

**Comment by tmtmtmtm** [comment]
@Abbie2020 — That's great, thanks! Shout if there's anything we haven't made clear on http://everypolitician.org/scrapers.html or if you've any questions

**Comment by Abbie2020** [comment]
I'm not sure about the terms of the MPs. I've only managed to find some data on the duration of each parliament: http://www.parliament.lk/en/duration-of-parliament

**Comment by Abbie2020** [comment]
https://en.wikipedia.org/wiki/15th_Parliament_of_Sri_Lanka#Members

**Comment by Abbie2020** [comment]
https://en.wikipedia.org/wiki/14th_Parliament_of_Sri_Lanka#List

**Comment by tmtmtmtm** [comment]
Those Wikipedia pages look very useful, but they imply there's a large amount of alliance-switching that goes on!

**Comment by Abbie2020** [comment]
Yeah, very odd! Should I scrape the Wikipedia pages for the term dates? There's no guarantee that the MP names will be the same as those on the Parliament website where all the other MP data is though...

**Comment by Abbie2020** [comment]
I've had some issues with the regex for getting the electoral district out. The name of the district appears to be an inconsistent number of spaces after 'Electoral District / National List' within the `<td>` cell tags, e.g. http://www.parliament.lk/en/members-of-parliament/directory-of-members/viewMember/53

```
<td ><div class="row_con">Electoral District / National List</div> 
                                Matara</td>
```

Can anyone help with this?

**Comment by tmtmtmtm** [comment]
> Should I scrape the Wikipedia pages for the term dates?

If you're just doing a single term or a small number, it's often quicker and easier to just build the term info by hand, rather than scraping it. If it's a lot, then scraping it might be quicker.

> There's no guarantee that the MP names will be the same as those on the Parliament website where all the other MP data is though

What we tend to do is write separate scrapers for each source. Then we already have a process for bringing those into EveryPolitician and merging, with thresholds for approximate matches etc.

See, for example, how we did Costa Rica: https://github.com/everypolitician/everypolitician-data/blob/master/data/Costa_Rica/Assembly/sources/instructions.json

The first two sources are separate scrapers for the official site, which match each other with a shared ID. Then we match to Wikipedia by name, as an approximate match. And then we pull in information Wikidata too, keyed on the Wikipedia names.

So, basically, if you scrape as many sources as you can find, we can take care of tying them all together :)

**Comment by tmtmtmtm** [comment]
> I've had some issues with the regex for getting the electoral district out.

What language are you writing your scraper in? In general doing this with a regex isn't recommended — you're much better using CSS or XPath selectors

So in Ruby, using Nokogiri, for example, you could say:

```
noko.xpath('.//div[@class="row_con" and contains(.,"Electoral District")]/following-sibling::text()').text.strip
```

**Comment by Abbie2020** [comment]
Is this format ok? I've been using Kimono.

Members of the 15th Parliament of Sri Lanka Wikipedia: https://www.kimonolabs.com/api/ondemand/d2dfhjjc

Additional scrapers to come.

**Comment by Abbie2020** [comment]
Members of the 14th Parliament of Sri Lanka Wikipedia:
http://www.kimonolabs.com/api/ondemand/e01y0hay

And the official Parliament of Sri Lanka pages (currently still scraping): http://www.kimonolabs.com/api/ondemand/4rqbmp8m

**Comment by tmtmtmtm** [comment]
Hi @Abbie2020 

Those links aren't working for me: they're reporting

```
{
  "status": "429 Too many requests",
  "error": "Based on your high request volume you are limited to one request per 60 seconds. If you need to make a high volume of requests to a single API it is recommended to use our crawling feature and provide a list of URLs directly."
}
```

Is the source public somewhere?

**Comment by Abbie2020** [comment]
You should be able to access https://www.kimonolabs.com/apis/4rqbmp8m if you create an account in Kimono but that's not ideal... Looks like I might need to quickly learn a different way to scrape!

**Comment by tmtmtmtm** [comment]
Hi @Abbie2020 — yeah, I think it would be much better if this was on Morph.io like most of our others (more info on that at http://everypolitician.org/scrapers.html), as it doesn't really look like we can use it from kimono easily at all.

I'd be happy to help you move your existing work over to there if you'd like.

**Comment by Abbie2020** [comment]
https://morph.io/Abbie2020/sri_lanka_members_of_parliament

Will try and do the Wikipedia pages too.

**Comment by tmtmtmtm** [comment]
Excellent, thanks!

Would you mind reversing the dates to ISO 8601 format (i.e. 1956-08-29 instead of 29-08-1956)?

It would also be very helpful if you could split the party initials out to a `party_id` column — e.g. instead of a `party` of "United National Party(UNP)" you would have a `party` of "United National Party", and a `party_id` of "UNP"

And instead of a string of "NULL" for things like missing email addresses it would be better if those were just blank.

But generally looking great, thanks!

**Comment by Abbie2020** [comment]
Take a look now :-)

**Comment by tmtmtmtm** [comment]
Yay, excellent! Thanks!

One other small thing (entirely my fault as this isn't documented anywhere — though it [should be!](https://github.com/everypolitician/viewer-sinatra/issues/683)) — it would be great if each row had a `source` column with the URL for the person.

I've also had a look at your code, and I'm a little surprised/confused as to why you've hard-coded all the IDs into it. Ideally it would get those from the list page, so it could run regularly and update automatically when people change etc. Is it difficult to make it do that?

**Comment by andylolz** [comment]
This looks v close.

Regarding the IDs: In case it helps, it looks like you can do something like this:

```python
index_url = "http://www.parliament.lk/members-of-parliament/index2.php?option=com_members&task=all&letter="
for letter in string.ascii_uppercase:
    url_to_fetch = index_url + letter
    # do stuff
```

**Comment by Abbie2020** [comment]
Sure, I'll do that. I realise it's anti-scraper to hardcode the IDs. I was trying to use XPath to get all the URLs from each page but it wasn't working and I got fed up with it! What @andylolz has posted looks like it'll do the job. Thanks Andy for posting it, is that definitely just current members and not past members as well?

**Comment by andylolz** [comment]
> is that definitely just current members and not past members as well?

Oh – v good question. I’ve no idea I’m afraid… _checks…_

**Comment by Abbie2020** [comment]
Thanks guys. Have another look now. By the way, if you have any other pointers about my code, I'd be grateful to know because I'm just teaching myself all this and learning as I go.

@andylolz How did you find the page with the IDs on? I think it'd be pretty easy to copy the code for the past members because the pages are almost exactly the same.

**Comment by andylolz** [comment]
[Huge apologies if I’m explaining stuff you already know!]

So I opened up [Developer tools in Chrome](https://developer.chrome.com/devtools), clicked the Network tab to show all requests made, then on the [directory of members page](http://www.parliament.lk/en/members-of-parliament/directory-of-members/?cletter=A) I clicked on one of the alphabet letters in the nav at the top. A request appears in dev tools, and clicking into it reveals the request URL. I pasted that URL into the browser, and just then just to tidy it up a bit I tried removing URL parameters until it stopped working.

Hope that helps! I think you’re right that you can do something similar for past members.

**Comment by andylolz** [comment]
> if you have any other pointers about my code

Ooh – one thing… Currently you check the status code of the response, but [fail silently if it isn’t 200](https://github.com/Abbie2020/sri_lanka_members_of_parliament/blob/4dbcad12/scraper.py#L76). The danger here is the scraper may appear to have succeeded, but has in fact stored an incomplete set of output data. I’d be tempted to either:

 * ignore the status code, and just fail if any request fails. Not ideal, but at least when the scraper succeeds you know the output data is correct
 * handle the error i.e. retry when you see a non-200 status code

If you’re seeing a lot of these non-200 status codes, my suggestion in the previous comment (adding `sleep`s in between requests) _might_ help reduce these.

**Comment by Abbie2020** [comment]
@tmtmtmtm What's your preferred method of error handling in these scrapers?

**Comment by tmtmtmtm** [comment]
Generally I go with the first of @andylolz's suggestions — just blow up if something goes wrong. That way I know whether the run was OK, and can be imported. If I were dealing with a site that had lots of problems on a recurring basis, I'd probably do something more complex, but just for getting the data in in the first place, then I'd say simpler is better!

**Comment by Abbie2020** [comment]
OK, I think I'm done then. I've added in a sleep between requests and removed the check for status code.

**Comment by tmtmtmtm** [comment]
Yay! Thanks @Abbie2020 — this looks great, and is now live at http://everypolitician.org/sri-lanka/

I've opened new tickets for the Wikipedia data (#888) and historic info (#889). Are you still thinking of trying those too?

**Comment by andylolz** [comment]
:rocket: :rocket: :rocket:

**Comment by Abbie2020** [comment]
Thanks for your help and for bearing with me :smile: 

I think the historic info should almost be copy and paste (fingers crossed!) so I'll give that a try.
